Add star rating / BPM / difficulty display while mod select is open

Givikap120 opened the pull request:
Video demonstration (1.2x UI scale)

https://github.com/ppy/osu/assets/89256026/b4e50f2b-a73e-45d3-a39d-a868fb6fe80b

Known issues:
- doesn't work on UI scale >1.2x (want to listen right solution idea)

Givikap120 committed:
Basic stuff (not working for now)

Givikap120 committed:
somewhat working prototype

Givikap120 committed:
Merge branch 'ppy:master' into map_info_on_mod_settings

Givikap120 committed:
functionality is done

Givikap120 committed:
Merge branch 'map_info_on_mod_settings' of https://github.com/Givikap120/osu into map_info_on_mod_settings

Givikap120 committed:
minor design fixes

Givikap120 committed:
Update AdvancedStats.cs

Givikap120 committed:
Update VerticalAttributeDisplay.cs

Givikap120 commented:
I've got an idea how to deal with UI scaling.
Difficulty attributes will be hidden and left click will make them appear

In "only SR and BPM" mode it should be enough space for pretty high UI scaling values (not sure about 1.6x)
And "opening" full info will shrink mod control buttons away

But this needs future discussion

Givikap120 committed:
Fixed "no animations" issue

Nabile-Rahmani commented:
> I've got an idea how to deal with UI scaling. Difficulty attributes will be hidden and left click will make them appear
> 
> In "only SR and BPM" mode it should be enough space for pretty high UI scaling values (not sure about 1.6x) And "opening" full info will shrink mod control buttons away
> 
> But this needs future discussion

Heh, I didn't read this comment before jumping in the image editor.

The difficulty container could be a collapsible (clickable) container.

When its parent container (or the collapsible container itself ?) detects that there is no more room to fit its content, the collapsible elements can automatically collapse themselves into a smaller form factor.

On normal UI scaling, this means the full info is always visible.

On larger scaling, you'd have to open the button to see details, and it could overlap the content. Not sure if easily doable with the existing containers.

Additionally, there's a lot of left padding that could automatically be reduced to fit more content.

I feel there's already code to do that as can be seen in `OsuTabControl > OsuTabDropdown > FillFlowContainer`, used on the song select sorting tab bar.

## Closed

![2x-closed](https://github.com/ppy/osu/assets/15223353/fdc85245-b67b-457c-9eab-df8d8da493e4)

## Opened

![2x-opened](https://github.com/ppy/osu/assets/15223353/d5323b9e-3a76-4566-bd0f-c02011d342bd)

![2x-opened-padding-note](https://github.com/ppy/osu/assets/15223353/bc7f1687-765b-455c-a471-f8551bd29997)

![2x-opened-no-padding](https://github.com/ppy/osu/assets/15223353/c6400ca4-9c33-4103-8d53-60db9d701b6b)

peppy commented:
I'm going to add this to the project as I think it's quite important to have visualised.

bdach commented on osu.Game/Beatmaps/BeatmapShortInfo.cs:
```diff
@@ -0,0 +1,29 @@
+
+using System;
+
+namespace osu.Game.Beatmaps
+{
+    public class BeatmapShortInfo : IEquatable<BeatmapShortInfo>
```
Cannot exist with this name and in this namespace. Should be a structure local to the mod overlay.

Probably something like `osu.Game.Overlays.Mods.ModEffectPreviewInfo` or something.

Also, for bonus points, if you make this a `struct` or a `record` (whichever is more convenient), you can skip the `Equals()` implementation.

Givikap120 replied:
deleted this completely

bdach commented on osu.Game/Overlays/Mods/ModMapInfoContainer.cs:
```diff
@@ -0,0 +1,184 @@
+            const float border_thickness = 2;
+
+            AutoSizeAxes = Axes.Both;
+            InternalChild = content = new InputBlockingContainer
```
This makes zero sense. The class inherits `Container`, but is overwriting the content `Container` _already_ provides it, but also not overriding `Container.Content`.

This should inherit `CompositeDrawable` instead.

Givikap120 replied:
changed to `CompositeDrawable`

bdach commented on osu.Game/Overlays/Mods/ModMapInfoContainer.cs:
```diff
@@ -0,0 +1,184 @@
+                Origin = Anchor.BottomRight,
+                Anchor = Anchor.BottomRight,
+                AutoSizeAxes = Axes.X,
+                Height = 50, // as ModSelectOverlay footer buttons
```
Extract `ShearedButton.Height` to a constant (`public const double HEIGHT = 50;`) and reference that instead.

Givikap120 replied:
fixed

bdach commented on osu.Game/Overlays/Mods/ModSelectOverlay.cs:
```diff
@@ -27,6 +27,7 @@
 using osu.Game.Rulesets.Mods;
 using osu.Game.Utils;
 using osuTK;
+using osuTK.Graphics;
```
```suggestion
```

Givikap120 replied:
fixed

bdach commented on osu.Game/Overlays/Mods/ModMapInfoContainer.cs:
```diff
@@ -0,0 +1,184 @@
+
+namespace osu.Game.Overlays.Mods
+{
+    public partial class ModMapInfoContainer : Container
```
Would suggest rename to `ModEffectPreviewPanel` or something

bdach commented on osu.Game/Screens/Select/BeatmapDetails.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System.Linq;
 using osu.Framework.Allocation;
+using osu.Framework.Bindables;
```
```suggestion
```

Givikap120 replied:
fixed

bdach commented on osu.Game/Screens/Select/Details/AdvancedStats.cs:
```diff
@@ -99,6 +102,32 @@ protected override void LoadComplete()
         private ScheduledDelegate debouncedStatisticsUpdate;
 
+        private StarDifficulty latestStarDifficulty = new StarDifficulty();
+        private void updateBindedInfo()
+        {
+            if (adjustedInfo == null) return;
+
+            // sadly need to calculate this to prevent additional data transportation
+            double rate = 1;
+            foreach (var mod in mods.Value.OfType<IApplicableToRate>())
+                rate = mod.ApplyToRate(0, rate);
+
+            double bpm = 0;
+            if (beatmapInfo != null) bpm = beatmapInfo.BPM * rate;
+
+            BeatmapShortInfo adjusted = new BeatmapShortInfo()
+            {
+                CircleSize = FirstValue.Value.adjustedValue ?? FirstValue.Value.baseValue,
+                DrainRate = HpDrain.Value.adjustedValue ?? HpDrain.Value.baseValue,
+                ApproachRate = ApproachRate.Value.adjustedValue ?? ApproachRate.Value.baseValue,
+                OverallDifficulty = Accuracy.Value.adjustedValue ?? Accuracy.Value.baseValue,
+                BPM = bpm,
+                StarDifficulty = latestStarDifficulty
+            };
+
+            adjustedInfo.Value = adjusted;
+        }
```
None of this logic can be here. To see why, perform the following exercise:

- Enter song select
- Switch active tab on the left to any leaderboard (local/global/friends/whatever)
- Open mod overlay
- The display will not update anymore, regardless of what you do

All of this should be implemented locally to the mod overlay. This also likely fixes issues like the tests failing because they don't have `Bindable<BeatmapShortInfo>` registered as a dependency or whatever.

Givikap120 replied:
Is calculating this locally gonna grab the Star Rating value from cache without calculating it twice?

Givikap120 replied:
fixed

Givikap120 committed:
fixed all stated problems

Givikap120 commented on osu.Game/Overlays/Mods/ModMapInfoContainer.cs:
```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
fixed

Givikap120 committed:
minor format fix

Givikap120 committed:
minor format fix (again)

Givikap120 committed:
Better counter formatization

Now CS and HP have 1 decimal point precision, while AR and OD is kept with 2 because of future support of rate-changed AR/OD

bdach committed:
Fix code inspections

bdach committed:
Extract constants properly

bdach committed:
Refactor preview panel to be more self-contained

bdach committed:
Add actual testing

bdach committed:
Refactor layout of display

bdach committed:
Fix data flow being sometimes incorrect

bdach committed:
Do not show mod effects on free mod select

bdach committed:
Merge branch 'master' into map_info_on_mod_settings

bdach committed:
Remove weird fade

bdach committed:
Fix slightly wrong binding setup

bdach committed:
Add collapsed state to effect preview panel

bdach committed:
Fix preview panel not being marked nullable

bdach committed:
Fix z-order of mod preview panel

bdach committed:
Collapse mod effect preview when it's not wide enough to fit

bdach committed:
Make mod effect preview uncollapse on hover

bdach commented:
As can probably be surmised from the above commit history, I've basically rewritten this. Highlights:

- Fixed weird interdependencies with `ModSelectOverlay` and made the component (sorta) self-contained
- Fixed the thing appearing on free mod select overlay which made no sense
- Made it collapse when there is not enough space to show it on screen
- Added actual testing

and other general refactorings and cleanups.

@Givikap120 please look through my changes and check how we generally expect things to be done. If there is anything unclear, please ask.
@ppy/team-client requesting second review because of the extent of changes made.

bdach committed:
Fix test failure

peppy committed:
Improve collapse/expand animations

Especially when on a screen resolution where it would start collapsed.

peppy committed:
Move shear spec to correct location

peppy committed:
Adjust various class naming and add some xmldoc

peppy committed:
Merge branch 'master' into map_info_on_mod_settings
